PuzzleTimeObjective-C/PuzzleTime/GameViewController.m

```diff
@@ -0,0 +1,382 @@
+                sourceIndexPath = indexPath;
+                PuzzleCell *cell = (PuzzleCell *)[_collectionView cellForItemAtIndexPath:sourceIndexPath];
+                
+                //Take a snapshot of the selected row using helper method
+                snapshot = [cell createSnapshot];
+                
+                //Add the snapshot as subview, centered at cell's center
+                __block CGPoint center = cell.center;
+                snapshot.center = center;
+                snapshot.alpha = 0.0;
+                [self.collectionView addSubview:snapshot];
+                [UIView animateWithDuration:0.25 animations:^{
+                    
+                    //Offset for gesture location.
+                    center = location;
+                    snapshot.center = center;
+                    snapshot.transform = CGAffineTransformMakeScale(1.05, 1.05);
+                    snapshot.alpha = 0.98;
+                    
+                    //Fade out
+                    cell.alpha = 0;
+                    
+                } completion:^(BOOL finished) {
+                }];
+            }
+            break;
+           
+        case UIGestureRecognizerStateChanged: {
+            CGPoint center = snapshot.center;
+            center = location;
+            snapshot.center = center;
+            
+            //Is destination valid and is it different from source?
+            if (indexPath && ![indexPath isEqual:sourceIndexPath]) {
+                
+                //update collectionView
+                [self swapTwoItemsAndSaveStatus:indexPath secondIndexPath:sourceIndexPath];
+                sourceIndexPath = indexPath;
+            }
+            break;
+        }
+            
+        default: {
+            //Clean up
+            PuzzleCell *cell = (PuzzleCell *)[_collectionView cellForItemAtIndexPath:sourceIndexPath];
+            cell.hidden = NO;
+            cell.alpha = 0;
+            [UIView animateWithDuration:0.25 animations:^{
+                
+                snapshot.center = cell.center;
+                snapshot.transform = CGAffineTransformIdentity;
+                snapshot.alpha = 0;
+                
+                //Undo fade out
+                cell.alpha = 1.0;
+                
+            } completion:^(BOOL finished) {
+                
+                sourceIndexPath = nil;
+                [snapshot removeFromSuperview];
+                snapshot = nil;
+            }];
+            break;
+        }
+            
+    }
+
+}
+
+- (void) swapTwoItemsAndSaveStatus: (NSIndexPath *) firstIndexPath secondIndexPath: (NSIndexPath *) secondeIndexPath {
+    NSLog(@"Swap first IndexPath: %ld, second IndexPath: %ld", (long)firstIndexPath.item, (long)secondeIndexPath.item);
+    NSArray *changedIndices = [NSArray arrayWithObjects:firstIndexPath, secondeIndexPath, nil];
+    [_curItemsArray exchangeObjectAtIndex:firstIndexPath.item withObjectAtIndex:secondeIndexPath.item];
```

Comment:
It would probably be better for `_curItemsArray` to live inside a `Game` model class that knows how to maintain/modify the game state. That will help remove some code from the controller.